Add custom Prisma serialization only when the schema is not empty

cprecioso opened the pull request:
Resolves #2798

`@prisma/client`, as installed from `npm` is just a placeholder empty client. When calling `prisma generate`, that placeholder gets overwritten with the actual client.

The problem is the placeholder is CJS and the actual built client is ESM. Together with [our workaround](https://github.com/wasp-lang/wasp/blob/a771c50e5cadc589d7fbd54b3fed0c0ec38545b4/waspc/data/Generator/templates/react-app/vite.config.ts#L45-L53) for correct Vite resolution of Prisma in our old version; Vite gets confused.

I made the Prisma importing conditional on there being actual entities in our AppSpec, so we will never import an empty client.

We can remove the workarounds when we:
- #2504 
- #2867

infomiho commented on waspc/data/Generator/templates/sdk/wasp/core/serialization.ts:
```diff
@@ -47,6 +54,13 @@ export interface SuperJSONObject {
 }
 
+{=# entitiesExist =}
+/*
+  This section is only included if there are (Prisma) entities in the app spec,
+  as it needs a non-empty Prisma client to work.
+  TODO: Check if we can remove this  conditional when we upgrade Prisma (#2504)
+*/
```
You can replace this comment with Mustache comment:
```
  {=! This section is only included if there are (Prisma) entities in the app spec, =}
```

This way this comment stays "internal" and doesn't get in the user's code.

infomiho commented on waspc/src/Wasp/Generator/SdkGenerator.hs:
```diff
@@ -243,6 +243,16 @@ depsRequiredForTesting =
+      object
+        [ "entitiesExist" .= entitiesExist
+        ]
+    entitiesExist = not . null $ getEntities spec
```
I'm tempted to ask you to extract this into a helper function that is named `arePrismaModelDefined` or `isPrismaUsed`. We have the same check repeated 4, 5 times now.

cprecioso commented:
@infomiho ready

cprecioso commented on waspc/data/Generator/templates/sdk/wasp/core/serialization.ts:
```diff
@@ -1,4 +1,9 @@
+{{={= =}=}}
```
I can do it, I had the same idea and was already thinking about this

cprecioso commented:
@infomiho ready

I split `serialization` into three files `base`, `prisma`, and `index`.

I moved all Prisma-specific things into the `prisma` file (only written if there are entities), and kept the rest in the `index`. `index` only has a conditional import of `./prisma`.

Given that registering custom types in superjson is a side-effect, I wanted to do it also for types. I copied a pattern from [TanStack Router](https://github.com/TanStack/router/blob/v1.124.1/docs/router/framework/react/guide/creating-a-router.md#router-type-safety) where files can register themselves into an interface. This way we keep the types working well either way.

**With Prisma file**
![image](https://github.com/user-attachments/assets/7fea1062-6331-42e6-a013-833509292153)

**Without Prisma file**
![image](https://github.com/user-attachments/assets/79594283-b3be-4c9b-8734-26b3749e4164)

infomiho commented on waspc/data/Generator/templates/sdk/wasp/core/serialization/base.ts:
```diff
@@ -0,0 +1,19 @@
+  https://github.com/TanStack/router/blob/v1.124.1/docs/router/framework/react/guide/creating-a-router.md#router-type-safety
+*/
+declare module "superjson" {
+  interface WaspInternal_CustomSerialization_Register {}
```
```suggestion
  interface WaspInternal_CustomSerializableJSONValue_Register {}
```

To maybe make it more in line with existing naming?

cprecioso commented:
Fixed comments, renamed `base` to `custom-register` @infomiho

infomiho commented:
I'm okay with merging 👍
